ORM 6 / Search 6 integration

fax4ever opened the pull request.

fax4ever committed:
TODOCHECK Skip decimal scale integration test

Hibernate ORM 6 does not put the decimal scale on the column. We need to find another way to get it (e.g.: taking it from the metamodel).

yrodiere commented on integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/model/DefaultDecimalScaleMappingIT.java:
```diff
@@ -37,8 +37,8 @@ public void mapping() {
 				.field( "unscaled", BigDecimal.class, f -> f.defaultDecimalScale( 0 ) )
-				.field( "defaultScaled", BigDecimal.class, f -> f.defaultDecimalScale( 2 ) )
-				.field( "hsearch", BigDecimal.class, f -> f.defaultDecimalScale( 2 ).decimalScale( 7 ) )
+				.field( "defaultScaled", BigDecimal.class, f -> { } )
```
This is not really a fix. With the stub backend, it will work, granted. But with real backends, decimal fields *require* a decimal scale, so this will lead to a mapping error, whereas in ORM 5 it would have worked and defaulted to 2.

I suppose ORM 6 still provides a default scale, but handles it differently. If so, we need to get this scale from Hibernate ORM metadata and communicate it to the backend somehow, similarly to what we do in `org.hibernate.search.mapper.orm.mapping.impl.HibernateOrmMappingPropertiesMetadataContributor#collectScale`.

If that's not possible, then we need to either decide on a backwards-compatible default ourselves, or document the limitation somewhere.

In any case, just changing the test to accept incorrect behavior doesn't seem the right call. Let's keep the test failing for now, to remind us there is something to fix here?

fax4ever replied:
Sure, makes sense!

fax4ever committed:
TODOCHECK Avoid to reuse query instances

fax4ever committed:
TODOCHECK Skip 2nd lvl cache tests

ORM 6 has some issues doing the multiloading of 2nd lvl cached entities: sometimes some null is retrieved using the API.

fax4ever committed:
TODOCHECK Avoid to use implicit select-clause

fax4ever committed:
TODOCHECK Detect ORM 6 dynamic entities

yrodiere commented on mapper/orm/src/main/java/org/hibernate/search/mapper/orm/loading/impl/TypeQueryFactory.java:
```diff
@@ -17,14 +17,16 @@
 		EntityDomainType<?> typeOrNull = metamodel.entity( entityPersister.getEntityName() );
-		if ( typeOrNull != null ) {
+
+		if ( typeOrNull != null && !( typeOrNull.getExpressableJavaTypeDescriptor() instanceof DynamicModelJtd ) ) {
```
So if I understand this correctly, dynamic-map entities now DO have an `EntityDomainType`, but they still cannot be queried using the Criteria API?

How weird.

Can you please adjust this commit so that it works for both Hibernate ORM 5 and Hibernate ORM 6? Applying patch to the ORM 6-only version of our code will be a hassle, so I'd rather avoid it where we can, and here I think we can. For example you can check `entityPersister.getMappedClass().equals(Map.class)`; it's a bit of a hack, but I think this should work on both ORM 5 and 6?

fax4ever replied:
> So if I understand this correctly, dynamic-map entities now DO have an EntityDomainType, but they still cannot be queried using the Criteria API?

You do
> entityPersister.getMappedClass().equals(Map.class)

I think that it is not exact the same. But I see the point of making the same code working for both 5 and 6.

yrodiere commented on util/internal/integrationtest/mapper/orm/src/main/java/org/hibernate/search/util/impl/integrationtest/mapper/orm/ReusableOrmSetupHolder.java:
```diff
@@ -500,7 +500,11 @@ private static void clearEntityInstances(SessionFactory sessionFactory, Hibernat
 
 	private static Query<?> createSelectOrDeleteAllOfSpecificTypeQuery(EntityType<?> entityType, Session session, String prefix) {
-		StringBuilder builder = new StringBuilder( prefix ).append( "from " ).append( entityType.getName() ).append( " e" );
+		StringBuilder builder = ( prefix == null || prefix.trim().isEmpty() ) ?
```
Please use `org.hibernate.search.util.common.impl.StringHelper` when you can:

```suggestion
		StringBuilder builder = StringHelper.isEmpty( prefix ) ?
```

Though to be honest, I don't think the prefix is ever null here. The old code wouldn't have worked with a `null` prefix.

yrodiere replied:
I guess you disagree :)

fax4ever replied:
Actually I miss it :( My bad

yrodiere commented on integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/massindexing/MassIndexingCachingIT.java:
```diff
@@ -38,13 +38,16 @@
 import org.assertj.core.api.AbstractLongAssert;
 import org.assertj.core.api.SoftAssertions;
 
+@Ignore // ORM 6 has some issues doing the multiloading of 2nd lvl cached entities:
+		// sometimes some null is retrieved using the API.
```
```suggestion
@Ignore( "ORM 6 has some issues doing the multiloading of 2nd lvl cached entities:"
		+ " sometimes some null is retrieved using the API." )
```

yrodiere commented on integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/model/DefaultDecimalScaleMappingIT.java:
```diff
@@ -37,8 +37,8 @@ public void mapping() {
-				.field( "hsearch", BigDecimal.class, f -> f.defaultDecimalScale( 2 ).decimalScale( 7 ) )
+				.field( "defaultScaled", BigDecimal.class, f -> { } )
+				.field( "hsearch", BigDecimal.class, f -> f.decimalScale( 7 ) )
 				.field( "notDecimal", Integer.class )
```
Can you please remove the commit "TODOCHECK Align decimal scale ORM default" and replace it with a commit that simply adds `@Ignore( "ORM 6 does not provide a default scale in the column metadata for BigDecimal fields" )` to `DefaultDecimalScaleMappingIT`?

yrodiere reviewed (commented):
A few comments to address before we merge.

Also, if you start ignoring tests, you should probably use `@Category`, so that we can more easily execute tests against ORM 5 but not ORM 6, when we will have separate modules:

* Create an interface called `RequiresOrm5`, put it in a `testsupport.categories` package somewhere in the module that holds the ORM integration tests. Similarly to what we do with `org.hibernate.search.integrationtest.backend.elasticsearch.testsupport.categories.RequiresSingleModelDialectForMajorVersion`.
* Put `@Category(RequiresOrm5.class)` on every test you wish to disable
* Set `<excludedGroups>` in the maven-failsafe-plugin configuration of the ORM integration tests, similarly to [this](https://github.com/hibernate/hibernate-search/blob/c186de03ed0e5512258f2c221ad3c9e5a6fb9ea4/integrationtest/backend/elasticsearch/pom.xml#L84-L87).

fax4ever commented:
Thanks @yrodiere for the review. I'm doing the changes now.

fax4ever commented:
Thank you for the suggestions, great as usual.

yrodiere commented:
Merged, thanks
